[concurrency] Implement swift_task_runOnMainActor/hop to main actor thunk

gottesmm opened the pull request:
[concurrency] Implement swift_task_runOnMainActor.

This routine takes a synchronous non-throwing main actor isolated closure
without a result. If we are dynamically on the main actor, we just run the
closure synchronously. Otherwise, we run a new task on the main actor and call
the closure on that.

This builds on top of the previous commit by using
swift_task_isCurrentExecutor_v2 in the implementation of this function.

To backwards deploy this function on Darwin, I used some tricks from libdispatch
to validate that we are on the main queue.

gottesmm committed:
[concurrency] Add support for HopToMainActorIfNeededThunk.

It is behind the experimental flag GenerateForceToMainActorThunks.

gottesmm committed:
[concurrency] Add a new entrypoint: swift_task_isCurrentExecutorWithFlags.

This entrypoint is similar to swift_task_isCurrentExecutor except that it
provides an ABI level option flag that enables one to configure its behavior in
a backwards deployable manner via the option flag.

I used this to expose at the ABI level the ability to check the current executor
without crashing on failure, while preserving the current behavior of
swift_task_isCurrentExecutor (which crashes on failure).

I am going to use this to implement swift_task_runOnMainActor.

gottesmm committed:
[concurrency] Implement swift_task_runOnMainActor.

This routine takes a synchronous non-throwing main actor isolated closure
without a result. If we are dynamically on the main actor, we just run the
closure synchronously. Otherwise, we run a new task on the main actor and call
the closure on that.

This builds on top of the previous commit by using
swift_task_isCurrentExecutorWithFlags in the implementation of this function.

To backwards deploy this function on Darwin, I used some tricks from libdispatch
to validate that we are on the main queue.

gottesmm commented:
@swift-ci smoke test

gottesmm commented:
@swift-ci smoke test

gottesmm commented:
@swift-ci smoke test

gottesmm commented:
I ripped out the backwards compatibility stuff. There is more to it and we may need some more compiler functionality or a compatibility library.

gottesmm commented:
but this is just for a prototype, so it makes sense to just get it done.

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -1259,3 +1259,53 @@ internal func _runTaskForBridgedAsyncMethod(@_inheritActorContext _ body: __owne
+@_silgen_name("swift_task_isCurrentExecutor_v2")
+@usableFromInline
+internal func _taskIsCurrentExecutor(
+  executor: Builtin.Executor, flags: UInt64) -> Bool
```
Can we put this one immediately next to the existing overload please?

ktoso reviewed (approved):
Yeah this is nicer now, by reusing the existing infra -- now it'd be great if we can get an SPI to make this checking more reliable -- Thanks for reusing the existing infra 👍

ktoso reviewed (commented):
I would prefer we rename the `swift_task_isCurrentExecutor_v2` -> `swift_task_isCurrentExecutorWithFlags` though to follow existing conventions, if this were to be merged.

gottesmm committed:
[thunk-lowering] Add support for mangling identity thunks.

This is just again using identity thunks to show that the behavior works before
I add in hop to main actor.

gottesmm commented:
@swift-ci smoke test

gottesmm commented:
@swift-ci toolchain

gottesmm commented:
Just going to do hop to main actor thunk in this as well.

gottesmm commented:
@swift-ci smoke test

gottesmm commented:
@swift-ci toolchain

gottesmm commented:
@swift-ci build toolchain
